exporter/otlphttpexporter/otlp.go

```diff
@@ -235,3 +239,81 @@ func readResponse(resp *http.Response) *status.Status {
 
 	return respStatus
 }
+
+func handlePartialSuccessResponse(resp *http.Response, partialSuccessHandler partialSuccessHandler) error {
+	if resp.ContentLength == 0 {
+		return nil
+	}
+
+	maxRead := resp.ContentLength
+	needsResize := false
+	if maxRead == -1 || maxRead > maxHTTPResponseReadBytes {
+		maxRead = maxHTTPResponseReadBytes
+		needsResize = true
+	}
+	protoBytes := make([]byte, maxRead)
+	n, err := io.ReadFull(resp.Body, protoBytes)
+
+	// No bytes read and an EOF error indicates there is no body to read.
+	if n == 0 && (err == nil || errors.Is(err, io.EOF)) {
+		return nil
+	}
+
+	// io.ReadFull will return io.ErrorUnexpectedEOF if the Content-Length header
+	// wasn't set, since we will try to read past the length of the body. If this
+	// is the case, the body will still have the full message in it, so we want to
+	// ignore the error and parse the message.
+	if err != nil && !errors.Is(err, io.ErrUnexpectedEOF) {
+		return err
+	}
+
+	// The pdata unmarshaling methods check for the length of the slice
+	// when unmarshaling it, so we have to trim down the length to the
+	// actual size of the data.
+	if needsResize {
+		protoBytes = protoBytes[:n]
+	}
+
+	return partialSuccessHandler(protoBytes)
+}
+
+type partialSuccessHandler func(protoBytes []byte) error
+
+func tracesPartialSuccessHandler(protoBytes []byte) error {
+	exportResponse := ptraceotlp.NewExportResponse()
+	err := exportResponse.UnmarshalProto(protoBytes)
+	if err != nil {
+		return err
+	}
+	partialSuccess := exportResponse.PartialSuccess()
+	if !(partialSuccess.ErrorMessage() == "" && partialSuccess.RejectedSpans() == 0) {
+		return consumererror.NewPermanent(fmt.Errorf("OTLP partial success: %s (%d rejected)", partialSuccess.ErrorMessage(), partialSuccess.RejectedSpans()))
```

Comment:
The spec also says servers can use the partial success to convey warnings/suggestions, even when all spans/metrics/logs were ingested: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#partial-success

>Servers MAY also make use of the partial_success field to convey warnings/suggestions to clients even when the request was fully accepted. In such cases, the rejected_<signal> field MUST have a value of 0 and the error_message field MUST be non-empty.

With the current handling we would miss such cases right?

Reply:
The same applies to gRPC.

Reply:
Never mind, I get it now. You can mark it as resolved!